modules/addons4.js

```diff
@@ -200,34 +201,101 @@ ScriptAddon.prototype.findUpdates = function(aUpdateListener, aReason) {
 };
 
 ScriptAddon.prototype._handleRemoteUpdate = function(
-    aUpdateListener, aAvailable) {
+    aUpdateListener, aResult, aInfo) {
   function tryToCall(obj, methName) {
     if (obj && ('undefined' != typeof obj[methName])) {
       obj[methName].apply(obj, Array.prototype.slice.call(arguments, 2));
     }
   }
 
+  var stringBundle = Components
+      .classes["@mozilla.org/intl/stringbundle;1"]
+      .getService(Components.interfaces.nsIStringBundleService)
+      .createBundle("chrome://greasemonkey/locale/gm-addons.properties");
+
+  var _scriptUpdatesFailure = stringBundle.GetStringFromName(
+      "script.updated.failure");
+
   try {
-    if (aAvailable) {
-      // Purge any possible ScriptInstall cache.
-      if (this.id in ScriptInstallCache) {
-        delete ScriptInstallCache[this.id];
-      }
-      // Then create one with this newly found update info.
-      var scriptInstall = ScriptInstallFactoryByAddon(
-          this, this._script);
-      AddonManagerPrivate.callInstallListeners(
-          'onNewInstall', [], scriptInstall);
-      tryToCall(aUpdateListener, 'onUpdateAvailable', this, scriptInstall);
-    } else {
-      tryToCall(aUpdateListener, 'onNoUpdateAvailable', this);
+    switch (aResult) {
+      case 0:
+        // Purge any possible ScriptInstall cache.
+        if (this.id in ScriptInstallCache) {
+          delete ScriptInstallCache[this.id];
+        }
+        // Then create one with this newly found update info.
+        var scriptInstall = ScriptInstallFactoryByAddon(
+            this, this._script);
+        AddonManagerPrivate.callInstallListeners(
+            'onNewInstall', [], scriptInstall);
+        tryToCall(aUpdateListener, 'onUpdateAvailable', this, scriptInstall);
+        tryToCall(aUpdateListener, 'onUpdateFinished', this,
+            AddonManager.UPDATE_STATUS_NO_ERROR);
+        break;
+      case 1:      
+        if (aInfo.error) {
+          GM_util.logError(
+              _scriptUpdatesFailure +
+              " " + aInfo.name + " - \"" + aInfo.url + "\"" +
+              " = " + aInfo.error);
+        } else {
+          /*
+          dump(
+              _scriptUpdatesFailure +
+              " " + aInfo.name + " - \"" + aInfo.url + "\"");
+          */
+        }
+        tryToCall(aUpdateListener, 'onNoUpdateAvailable', this);
+        tryToCall(aUpdateListener, 'onUpdateFinished', this,
+            AddonManager.UPDATE_STATUS_DOWNLOAD_ERROR);
+        break;
+      case 2:      
+        GM_util.logError(
+            _scriptUpdatesFailure +
+            " " + aInfo.name + " - \"" + aInfo.url + "\" = " + "timeout");
+        tryToCall(aUpdateListener, 'onNoUpdateAvailable', this);
+        tryToCall(aUpdateListener, 'onUpdateFinished', this,
+            AddonManager.UPDATE_STATUS_TIMEOUT);
+        break;
+      case 3:      
+        var _error = _scriptUpdatesFailure +
```

Comment:
Most of these cases look almost identical to each other.  Does this really take 70 lines to write?